compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/XCLBinGen.cpp

```diff
@@ -1116,12 +1127,36 @@ static LogicalResult generateUnifiedObject(
       return failure();
     }
   }
-  copy->erase();
+  moduleWithOneDevice->erase();
   return success();
 }
 
+FailureOr<ArrayRef<uint32_t>> getNpuInstructions(AIE::DeviceOp deviceOp) {
+  MLIRContext *ctx = deviceOp.getContext();
+  mlir::Attribute maybeNpuInstructions = deviceOp->getAttr("npu_instructions");
```

Comment:
Sorry I had a long comment here that got erased/forgotten (so I'm gonna stamp this as another review just so you see it):

This is what I was thinking about when I said messing with deviceop inside of a moduleop that's nowhere to be found might be bad juju. Like I know that if you run an operationpass on an op then the strong (verbotten) recommendation is you do not touch ops above/outside of the op that the operationpass is scoped too. We're not running an operationpass right here but we're touching something that's ostensibly outside of the deviceop - the npu_instructions aren't actually stored in the deviceop (I don't think), they're stored as a resource but on the moduleop. Obviously, currently, we're not mutating/touching that resource anywhere but it just leaves the door open for a can of worms I think.

Reply:
Not sure I follow  `npu_instructions` are an attribute on a `DeviceOp`. `npu_instructions` originate as `amdaie.controlcode` (inside a workgroup), then become `aiex.runtime_sequence` inside `aie.device`, and then become attribute on `aie.device`. Never go into the parent scope of the `aie.device`. 

Or are you saying, the actual serialization under the hood puts all resources on the `ModuleOp`? I'd be surprised if that's the case.

Reply:
> Not sure I follow npu_instructions are an attribute on a DeviceOp.

Look at the IR - the attribute stores a name to data that is attached to the `ModuleOp`.

> Or are you saying, the actual serialization under the hood puts all resources on the `ModuleOp`? I'd be surprised if that's the case.

That's exactly what I'm saying

Reply:
see https://github.com/nod-ai/iree-amd-aie/blob/6a0c637c9da26022aa6f094e3496b6982b885c79/compiler/plugins/target/AMD-AIE/aie/test/aiert_insts.mlir#L10-L16

Reply:
Where's the module op there? I just see a map, not connected to any op, probably floating free in the MLIRContext

Reply:
Actually you're right, I though it was inside the module but here's the full IR

```mlir
module {
  aie.device(npu1_4col) {
    memref.global "public" @of_toMem : memref<32xi32>
    memref.global "public" @of_fromMem : memref<32xi32>
    aie.shim_dma_allocation @of_fromMem(MM2S, 0, 0)
    aie.shim_dma_allocation @of_toMem(S2MM, 0, 0)
  } {npu_instructions = dense_resource<npu_instructions> : tensor<64xui32>}
}

{-#
  dialect_resources: {
    builtin: {
      npu_instructions: "0x0400000000010306050100000600000000010000010000000000000020D0010030000000200000000000000000000000000000000000008000000000000000000000000281000000300000000000000000000000000000000000000024D001000000000002000000000000000000000000000000000000000000000004D20100000000000100008018000000010000000000000000D001003000000020000000800000000000000000008000070020800F000000000000000000000281000000300000000000000000000000000000000000000004D001000000000000000000000000008000000000000000000000000000000014D20100000000000000000018000000"
    }
  }
#-}
```

so I guess that mean it's stored/uniqued in the dialect itself. My bad (really). But it makes me wonder if I made a mistake using asmresource like this...

Reply:
Hmm, using resource attribute seems fine/good to me. Nice when printed to have the big npu_instructions string at the very end.